### PR TITLE
Player Panel human name reset and ban

### PR DESCRIPTION
**Pull request by morrowwolf**
## About The Pull Request

This adds a section to handle human names in the player panel punish panel. 
Human name reset will allow you to change the name manually or have a random name selected and will set the preference name correctly.
Human name ban will disallow people from using custom names but they can still randomize until they find something they like.

## Why It's Good For The Game

Admins no longer have to tell "BEST MARINE" to change their name preferences four rounds in a row.

## Changelog

:cl: Morrow
admin: Added human name reset and human name ban to player panel.
/:cl:

### CODE REVIEW

File: code/modules/admin/player_panel/actions/punish.dm [hunk 1]
```diff
@@ -152,3 +152,102 @@
...
+		var/raw_name = input(user, "Choose the new name:", "Name Input")  as text|null
+		if(!isnull(raw_name)) // Check to ensure that the user entered text (rather than cancel.)
+			new_name = reject_bad_name(raw_name)
+
+	if(!new_name)
+		to_chat(user, SPAN_NOTICE("Invalid name. The name should be at least 2 and at most [MAX_NAME_LEN] characters long. It may only contain the characters A-Z, a-z, -, ' and ."))
+		return
+
+	target_mob.change_real_name(target_mob, new_name)
+	if(istype(target_mob, /mob/living/carbon/human))
```
Comment (stanalbatross) [inline code/modules/admin/player_panel/actions/punish.dm:192]: ishuman

File: code/modules/admin/player_panel/actions/punish.dm [hunk 2]
```diff
@@ -152,3 +152,102 @@
...
+			target_human.wear_id.registered_name = "[target_human.real_name]"
+			if(target_human.wear_id.assignment)
+				target_human.wear_id.name += " ([target_human.wear_id.assignment])"
+
+	target_mob.client.prefs.real_name = new_name
+	target_mob.client.prefs.save_character()
+
+	message_staff("[user.ckey] has reset [target_mob.ckey] name")
+
+	to_chat(target_mob, SPAN_DANGER("Warning: Your name has been reset by [user.ckey]."))
```
Comment (stanalbatross) [inline code/modules/admin/player_panel/actions/punish.dm:205]: make this bigger and accompany with a sound

File: code/modules/admin/player_panel/actions/punish.dm [hunk 3]
```diff
@@ -152,3 +152,102 @@
...
+		if(target_human.wear_id)
+			target_human.wear_id.name = "[target_human.real_name]'s ID Card"
+			target_human.wear_id.registered_name = "[target_human.real_name]"
+			if(target_human.wear_id.assignment)
+				target_human.wear_id.name += " ([target_human.wear_id.assignment])"
+
+	target_mob.client.prefs.real_name = new_name
+	target_mob.client.prefs.save_character()
+
+	message_staff("[user.ckey] has reset [target_mob.ckey] name")
```
Comment (stanalbatross) [inline code/modules/admin/player_panel/actions/punish.dm:203]:
```suggestion
	message_staff("[user.ckey] has reset [target_mob.ckey]'s name.")
```

File: code/modules/admin/player_panel/actions/punish.dm [hunk 4]
```diff
@@ -152,3 +152,102 @@
 
 	var/mob/living/carbon/Xenomorph/X = target
 	X.generate_name()
+
+/datum/player_action/reset_human_name
+	action_tag = "reset_human_name"
+	name = "Reset Human Name"
+
+
+/datum/player_action/reset_human_name/act(var/client/user, var/mob/target, var/list/params)
```
Comment (stanalbatross) [inline code/modules/admin/player_panel/actions/punish.dm:161]: needs a check for if they have a client

File: code/modules/admin/player_panel/actions/punish.dm [hunk 5]
```diff
@@ -152,3 +152,102 @@
...
+
+/datum/player_action/ban_human_name/act(var/client/user, var/mob/target, var/list/params)
+	if(!target.client || !target.ckey)
+		to_chat(user, SPAN_NOTICE("Target is lacking either client or ckey. Aborting."))
+		return
+
+	var/client/target_client = target.client
+
+	if(target_client.human_name_ban)
+		if(alert(user, "Are you sure you want to UNBAN [target.ckey] and let them use human names?", ,"Yes", "No") == "No")
```
Comment (stanalbatross) [inline code/modules/admin/player_panel/actions/punish.dm:222]: null arg bad

alert()s can take a long time

requires another client check and ckey check after

File: code/modules/admin/player_panel/actions/punish.dm [hunk 6]
```diff
@@ -152,3 +152,102 @@
...
+			return
+		target_client.human_name_ban = FALSE
+		target_client.prefs.human_name_ban = FALSE
+
+		target_client.prefs.save_preferences()
+		message_staff("[user.ckey] has unbanned [target.ckey] from using human names.")
+
+		notes_add(target.ckey, "Human Name Unbanned by [user.ckey]", user.mob)
+
+		to_chat(target, SPAN_DANGER("Warning: You can use human names again."))
```
Comment (stanalbatross) [inline code/modules/admin/player_panel/actions/punish.dm:232]: bigger

File: code/modules/admin/player_panel/actions/punish.dm [hunk 7]
```diff
@@ -152,3 +152,102 @@
...
+
+
+	if(alert("Are you sure you want to BAN [target.ckey] from ever using any human names?", , "Yes", "No") == "No")
+		return
+
+	message_staff("[user.ckey] has banned [target.ckey] from using human names.")
+
+	notes_add(target.ckey, "Human Name Banned by [user.ckey]", user.mob)
+
+	to_chat(target, SPAN_DANGER("Warning: You were banned from using human names by [user.ckey]."))
```
Comment (stanalbatross) [inline code/modules/admin/player_panel/actions/punish.dm:243]: as above

File: code/modules/admin/player_panel/actions/punish.dm [hunk 8]
```diff
@@ -152,3 +152,102 @@
...
+
+		notes_add(target.ckey, "Human Name Unbanned by [user.ckey]", user.mob)
+
+		to_chat(target, SPAN_DANGER("Warning: You can use human names again."))
+		return
+
+
+	if(alert("Are you sure you want to BAN [target.ckey] from ever using any human names?", , "Yes", "No") == "No")
+		return
+
```
Comment (stanalbatross) [inline code/modules/admin/player_panel/actions/punish.dm:238]: alert()s can take a long time

requires another client check and ckey check after

File: code/modules/admin/player_panel/actions/punish.dm [hunk 9]
```diff
@@ -152,3 +152,102 @@
...
+	if(!target_mob.ckey)
+		to_chat(user, SPAN_DANGER("Warning: Mob ckey for [target_mob.name] not found."))
+		return
+
+	if(alert(user, "Are you sure you want to reset name for [target_mob.ckey]?", , "Yes", "No") == "No")
+		return
+
+	var/new_name
+
+	if(alert(user, "Do you want to manually set the name for [target_mob.ckey]?", , "Yes", "No") == "No")
```
Comment (stanalbatross) [inline code/modules/admin/player_panel/actions/punish.dm:180]: alert()s can take a long time

requires another client check and ckey check after

**Review by stanalbatross — CHANGES_REQUESTED** [review]

File: code/modules/admin/player_panel/actions/punish.dm [hunk 10]
```diff
@@ -152,3 +152,122 @@
...
+
+/datum/player_action/ban_human_name/act(var/client/user, var/mob/target, var/list/params)
+	if(!target.client || !target.ckey)
+		to_chat(user, SPAN_NOTICE("Target is lacking either client or ckey. Aborting."))
+		return
+
+	var/client/target_client = target.client
+
+	if(target_client.human_name_ban)
+		if(alert(user, "Are you sure you want to UNBAN [target.ckey] and let them use human names?", ,"Yes", "No") == "No")
```
Comment (stanalbatross) [inline code/modules/admin/player_panel/actions/punish.dm:232]: null args

**Review by stanalbatross — CHANGES_REQUESTED** [review]

File: tgui/packages/tgui/interfaces/PlayerPanel.js [hunk 11]
```diff
@@ -415,6 +415,30 @@ const PunishmentActions = (props, context) => {
...
+            onClick={() => act("reset_human_name")}
+          />
+          <Button
+            width="100%"
+            height="100%"
+            icon="clipboard-list"
+            color="bad"
+            content="Human name ban"
+            disabled={!hasPermission(data, "ban_human_name")}
+            onClick={() => act("ban_human_name")}
```
Comment (stanalbatross) [inline tgui/packages/tgui/interfaces/PlayerPanel.js:438]: if they are already unbanned the button should show that by clicking it you'll unban them

**Review by stanalbatross — CHANGES_REQUESTED** [review]

File: code/modules/admin/player_panel/player_panel.dm [hunk 12]
```diff
@@ -508,6 +508,8 @@
 		.["client_rank"] = targetClient.admin_holder ? targetClient.admin_holder.rank : "Player"
 		.["client_muted"] = targetClient.prefs.muted
 
+		.["client_name_banned_status"] = targetMob.client.human_name_ban
```
Comment (stanalbatross) [inline code/modules/admin/player_panel/player_panel.dm:511]:
```suggestion
		.["client_name_banned_status"] = targetClient.human_name_ban
```

File: code/modules/admin/player_panel/actions/punish.dm [hunk 13]
```diff
@@ -152,3 +152,123 @@
...
+	if(!ismob(target_mob))
+		to_chat(user, SPAN_WARNING("[target.name] is not a mob!"))
+		return
+
+	if(isXeno(target_mob))
+		to_chat(user, SPAN_WARNING("[target.name] is a xeno!"))
+		return
+
+	if(!target_mob.client)
+		to_chat(user, SPAN_WARNING("[target.name] does not have a client!"))
```
Comment (Geevies) [inline code/modules/admin/player_panel/actions/punish.dm:172]: return

File: code/modules/admin/player_panel/actions/punish.dm [hunk 14]
```diff
@@ -152,3 +152,123 @@
...
+	if(!target_mob.ckey)
+		to_chat(user, SPAN_DANGER("Warning: Mob ckey for [target_mob.name] not found."))
+		return
+
+	if(alert(user, "Are you sure you want to reset name for [target_mob.ckey]?", "Confirmation", "Yes", "No") == "No")
+		return
+
+	var/new_name
+
+	if(alert(user, "Do you want to manually set the name for [target_mob.ckey]?", "Confirmation", "Yes", "No") == "No")
```
Comment (Geevies) [inline code/modules/admin/player_panel/actions/punish.dm:183]: should add a third option here that lets em cancel out if they got this far and regret it

File: code/modules/admin/player_panel/actions/punish.dm [hunk 15]
```diff
@@ -152,3 +152,123 @@
...
+		var/raw_name = input(user, "Choose the new name:", "Name Input")  as text|null
+		if(!isnull(raw_name)) // Check to ensure that the user entered text (rather than cancel.)
+			new_name = reject_bad_name(raw_name)
+
+	if(!new_name)
+		to_chat(user, SPAN_NOTICE("Invalid name. The name should be at least 2 and at most [MAX_NAME_LEN] characters long. It may only contain the characters A-Z, a-z, -, ' and ."))
+		return
+
+	if(!target_mob.client)
+		to_chat(user, SPAN_WARNING("[target.name] does not have a client!"))
```
Comment (Geevies) [inline code/modules/admin/player_panel/actions/punish.dm:195]: return

**Review by Geevies — CHANGES_REQUESTED** [review]

**Pull request merged**